modules/apps/27-interchain-accounts/keeper/account.go

```diff
@@ -4,51 +4,40 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+
+	"github.com/cosmos/ibc-go/modules/apps/27-interchain-accounts/types"
 	channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
 	host "github.com/cosmos/ibc-go/modules/core/24-host"
-	"github.com/cosmos/ibc-go/modules/apps/27-interchain-accounts/types"
-	"github.com/tendermint/tendermint/crypto/tmhash"
 )
 
-// The first step in registering an interchain account
-// Binds a new port & calls OnChanOpenInit
+// InitInterchainAccount is the entry point to registering an interchain account.
+// It generates a new port identifier using the owner address, connection identifier,
+// and counterparty connection identifier. It will bind to the port identifier and
+// call 04-channel 'ChanOpenInit'. An error is returned if the port identifier is
+// already in use. Gaining access to interchain accounts whose channels have closed
+// cannot be done with this function. A regular MsgChanOpenInit must be used.
 func (k Keeper) InitInterchainAccount(ctx sdk.Context, connectionId, owner string) error {
 	portId := k.GeneratePortId(owner, connectionId)
 
-	// Check if the port is already bound
-	isBound := k.IsBound(ctx, portId)
-	if isBound == true {
+	// check if the port is already bound
+	if k.IsBound(ctx, portId) {
 		return sdkerrors.Wrap(types.ErrPortAlreadyBound, portId)
 	}
 
 	portCap := k.portKeeper.BindPort(ctx, portId)
 	err := k.ClaimCapability(ctx, portCap, host.PortPath(portId))
 	if err != nil {
+		return sdkerrors.Wrap(err, "unable to bind to newly generated portID")
+	}
+
+	msg := channeltypes.NewMsgChannelOpenInit(portId, types.Version, channeltypes.ORDERED, []string{connectionId}, types.PortID, types.ModuleName)
+	handler := k.msgRouter.Handler(msg)
+	if _, err := handler(ctx, msg); err != nil {
```

Comment:
Just out of interest, why did you go with this approach instead of what was there? It certainly looks cleaner with the refactor, is that the only reason?

Reply:
No strong reasons. The previous functionality was definitively not ideal (calling 04-channel keeper and app callback). A better fix assumes that the 04-channel function handles all necessary functionality including app callback. Any change to this contract could cause future issues. Since we already have the message router, it seemed like the best option to me. It also aligns with the direction the SDK is pushing for so I figured future breaking changes would be easier to handle using this functionality